### PR TITLE
Improved robot state publisher

### PR DESCRIPTION
**Pull request by lrapetti**
Following the suggestion of @traversaro in https://github.com/robotology/idyntree/pull/513#issuecomment-493935969 and https://github.com/robotology/idyntree/pull/535#issuecomment-493935618, I am opening this PR from the branch in my fork that was also rebased on `devel`.

This PR deprecates https://github.com/robotology/idyntree/pull/513#issuecomment-493935969 and https://github.com/robotology/idyntree/pull/535#issuecomment-493935618.

CC @prashanthr05 @Yeshasvitvs

### CODE REVIEW

**Comment by lrapetti** [comment]
As observed in https://github.com/robotology/idyntree/pull/513#issuecomment-493929289, proper documentation is still missing, but I would proceed with merging for the time being.

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp [hunk 1]
```diff
@@ -62,24 +62,36 @@ YARPRobotStatePublisherModule::YARPRobotStatePublisherModule(): m_iframetrans(nu
...
+
     pTransformclient_cfg.put("remote", "/transformServer");
 
+    m_tfPrefix = rf.check("tfPrefix",Value("")).asString();
+
     bool ok_client = m_ddtransformclient.open(pTransformclient_cfg);
     if (!ok_client)
     {
         yError()<<"Problem in opening the transformClient device";
+        yError()<<"Is the transform server running?";
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]:
```suggestion
        yError()<<"Is the transformServer YARP device running?";
```

File: src/tools/yarprobotstatepublisher/src/main.cpp [hunk 2]
```diff
@@ -35,9 +35,11 @@ int main(int argc, char *argv[])
     if (rf.check("help"))
     {
         cout<<"Options"<<endl;
-        cout<<"\t--robot                <robot-name>: robot name"<<endl;
+        cout<<"\t--namePrefix             <name-prefix>: prefix of the yarprobotstatepublisher ports"<<endl;
+        cout<<"\t--tfPrefix               <tf-prefix>: prefix of the published TFs"<<endl;
         cout<<"\t--model                  <file-name>: file name of the model to load at startup"<<endl;
         cout<<"\t--base-frame             <frame-name>: specify the base frame of the published tf tree"<<endl;
+        cout<<"\t--jointstates-topic      <topic-name>: source topic that streams the joint state"<<endl;
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/src/main.cpp]: The default value of `jointstates-topic`  should be documented in  this help.

File: src/tools/yarprobotstatepublisher/src/main.cpp [hunk 3]
```diff
@@ -35,9 +35,11 @@ int main(int argc, char *argv[])
     if (rf.check("help"))
     {
         cout<<"Options"<<endl;
-        cout<<"\t--robot                <robot-name>: robot name"<<endl;
+        cout<<"\t--namePrefix             <name-prefix>: prefix of the yarprobotstatepublisher ports"<<endl;
+        cout<<"\t--tfPrefix               <tf-prefix>: prefix of the published TFs"<<endl;
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/src/main.cpp]: The default value of `tfPrefix`  should be documented in  this help.

File: src/tools/yarprobotstatepublisher/src/main.cpp [hunk 4]
```diff
@@ -35,9 +35,11 @@ int main(int argc, char *argv[])
     if (rf.check("help"))
     {
         cout<<"Options"<<endl;
-        cout<<"\t--robot                <robot-name>: robot name"<<endl;
+        cout<<"\t--namePrefix             <name-prefix>: prefix of the yarprobotstatepublisher ports"<<endl;
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/src/main.cpp]: The style of the existing options seems to be dash, separated, so I guess for consistency it would make sense to have all these options to be `--name-prefix` (the docs indeed already refer to them as that).

File: src/tools/yarprobotstatepublisher/src/main.cpp (hunk 2 above)
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/src/main.cpp]:
```suggestion
        cout<<"\t--jointstates-topic      <topic-name>: source ROS topic that streams the joint state"<<endl;
```

File: src/tools/yarprobotstatepublisher/src/main.cpp (hunk 4 above)
Comment (traversaro) [reply src/tools/yarprobotstatepublisher/src/main.cpp]: By the way, "name-prefix" is a bit ambiguous as an option. If we want to change from the even more ambigous `robot`, what about `robot-port-prefix` or something similar?

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp [hunk 5]
```diff
@@ -209,9 +217,18 @@ void YARPRobotStatePublisherModule::onRead(JointState &v)
...
     {
-        iDynTree::JointIndex jntIndex = model.getJointIndex(v.name[i]);
+        for (size_t j=0; j < model.getNrOfDOFs(); j++)
+        {
+            if (v.name[i] == model.getJointName(j))
+            {
+                jntIndex = model.getJointIndex(v.name[i]);
+                break;
+            }
+        }
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: If am quite confused by this logic. What is the rationale for that? If you want to make sure that a given joint in the iDynTree models has a number of dofs that are equal to 1, you can just call, 
`model.getJoint(jntIndex)->getNrOfDOFs()` and check if it is equal to 1.

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp [hunk 6]
```diff
@@ -222,16 +239,50 @@ void YARPRobotStatePublisherModule::onRead(JointState &v)
...
-        if(m_baseFrameIndex == frameIdx)    // skip self-tranform
-            continue;
-
-        iDynTree::Transform base_H_frame = m_kinDynComp.getRelativeTransform(m_baseFrameIndex, frameIdx);
-        iDynTree::toYarp(base_H_frame.asHomogeneousTransform(), m_buf4x4);
-        m_iframetrans->setTransform(model.getFrameName(frameIdx),
-                                    model.getFrameName(m_baseFrameIndex),
-                                    m_buf4x4);
+    bool m_publishGlobalTF = false; // TODO
+    if (m_publishGlobalTF) {
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: I am bit confused by this if. What is this about? If the `if` is never executed, why is this included?

File: src/tools/yarprobotstatepublisher/src/main.cpp [hunk 7]
```diff
@@ -35,9 +35,11 @@ int main(int argc, char *argv[])
     if (rf.check("help"))
     {
         cout<<"Options"<<endl;
-        cout<<"\t--robot                <robot-name>: robot name"<<endl;
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/src/main.cpp]: `yarprobotstatepublisher` is a relatively stable tool, that is used by several users of iDynTree, so we can't simply change the name of an option, but we should first deprecated it (by still supporting it in addition to the new option) but printing a warning message, and then remove it after one stable release (so this change will be released in 0.12, and we can remove it for the 0.13 release). See https://github.com/robotology/yarp/blob/master/.github/CONTRIBUTING.md#deprecated-features for some related docs.

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp (hunk 6 above)
Comment (lrapetti) [reply src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: Actually, this dates back to the original WIP commit by @diegoferigo (https://github.com/lrapetti/idyntree/commit/6aea5f92a83cd66d8bd0176aedc1229c30e0cb73). If it can be a useful feature, we can add the option as a possible configuration.

Comment (traversaro) [reply src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: @diegoferigo Any comment. 
@lrapetti If you are not using this option, I suggest to just remove it from the PR. We can always add it later.

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp (hunk 5 above)
Comment (prashanthr05) [reply src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: Here, I guess we are checking the consistency between the JointState message joints and the Model joints.

Aside: I'd suggest to be more verbose with the variable names to make it more readable.

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp (hunk 6 above)
Comment (diegoferigo) [reply src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: I have no memory about this, probably at the time I wanted to make an option in a later commit (this one was still marked as WIP).

Comment (traversaro) [reply src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: Ack, given that it is always disabled I think it is not necessary for the use with HDE.

File: src/tools/yarprobotstatepublisher/src/main.cpp (hunk 2 above)
Comment (lrapetti) [reply src/tools/yarprobotstatepublisher/src/main.cpp]: done https://github.com/robotology/idyntree/pull/538/commits/2014337cc5eebe96c22e23f33a5948e4d3c62f8d

File: src/tools/yarprobotstatepublisher/src/main.cpp (hunk 3 above)
Comment (lrapetti) [reply src/tools/yarprobotstatepublisher/src/main.cpp]: https://github.com/robotology/idyntree/pull/538/commits/2014337cc5eebe96c22e23f33a5948e4d3c62f8d

File: src/tools/yarprobotstatepublisher/src/main.cpp (hunk 4 above)
Comment (lrapetti) [reply src/tools/yarprobotstatepublisher/src/main.cpp]: > The style of the existing options seems to be dash, separated, so I guess for consistency it would make sense to have all these options to be `--name-prefix` (the docs indeed already refer to them as that).

Done.

> what about `robot-port-prefix` or something similar?

It is the prefix for all the ports opened by the `yarprobotstatepusblisher` (`/yarprobotstatepublisher`, `/yarprobotstatepublisher/transformClient/rpc:i`, `/yarprobotstatepublisher/transformClient/rpc:o`, `yarprobotstatepublisher/transformClient/transforms:i`), I don't see why to add the word `robot` to this option. However, if you think it can give more continuity to the previous name (`robot`) for me its fine.

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp (hunk 6 above)
Comment (lrapetti) [reply src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: Done https://github.com/robotology/idyntree/pull/538/commits/3625d301a81976521c273c9574489553da381d43

File: src/tools/yarprobotstatepublisher/CMakeLists.txt [hunk 8]
```diff
@@ -1,17 +1,22 @@
...
+    src/main.cpp
+    src/robotstatepublisher.cpp)
+
+target_include_directories(${PROJECT_NAME} PRIVATE ${PROJECT_SOURCE_DIR}/include)
+
+target_link_libraries(${PROJECT_NAME}
+    ${YARP_LIBRARIES}
+    YARP::YARP_rosmsg
+    idyntree-high-level
+    idyntree-yarp)
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/CMakeLists.txt]: I am not following everything in this changes (a typical problem in big PRs), but I guess that now we can remove the file https://github.com/robotology/idyntree/blob/master/src/tools/yarprobotstatepublisher/JointState.msg .

Comment (lrapetti) [reply src/tools/yarprobotstatepublisher/CMakeLists.txt]: I agree on that. Done https://github.com/robotology/idyntree/pull/538/commits/bdb55caaa8689852faa5f172236b681cc42b31a7

File: src/tools/yarprobotstatepublisher/src/main.cpp (hunk 7 above)
Comment (lrapetti) [reply src/tools/yarprobotstatepublisher/src/main.cpp]: I don't see any standard message for deprecated options, so I am just printing the warning with `std::cerr` (https://github.com/robotology/idyntree/pull/538/commits/a29e145bf2859d7da46dd479339b5bbcd557e26f). 

@traversaro let me know if there is a prefered way to do that.

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp (hunk 5 above)
Comment (lrapetti) [reply src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: > I guess we are checking the consistency between the JointState message joints and the Model joints

If a joint name in the jointState messag is not found in the model, `getJointIndex` return `JOINT_INVALID_INDEX ` which was already taken into account in the following check (https://github.com/robotology/idyntree/pull/538/files/8cd4a1b3cd8a1ada7859cb859eb31dc04310f528#diff-c2170ce79e08829b7556ba58dcbca3a3L215).

Honestly, I am a bit lost too in understanding the purpose here. For how I see it, we don't need this check.

Comment (traversaro) [reply src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: > Honestly, I am a bit lost too in understanding the purpose here. For how I see it, we don't need this check.

Then I think we can remove it.

File: src/tools/yarprobotstatepublisher/src/main.cpp (hunk 7 above)
Comment (traversaro) [reply src/tools/yarprobotstatepublisher/src/main.cpp]: Printing a simple warning with `std::cerr` is fine.

Comment (lrapetti) [reply src/tools/yarprobotstatepublisher/src/main.cpp]: > Printing a simple warning with `std::cerr` is fine.

ok, then its done in https://github.com/robotology/idyntree/commit/a29e145bf2859d7da46dd479339b5bbcd557e26f

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp (hunk 5 above)
Comment (lrapetti) [reply src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: Done https://github.com/robotology/idyntree/pull/538/commits/c15230ac792c72523668abc4506510d862748224

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp [hunk 9]
```diff
@@ -229,14 +229,7 @@ void YARPRobotStatePublisherModule::onRead(yarp::rosmsg::sensor_msgs::JointState
...
     {
-        for (size_t j=0; j < model.getNrOfDOFs(); j++)
-        {
-            if (v.name[i] == model.getJointName(j))
-            {
-                jntIndex = model.getJointIndex(v.name[i]);
-                break;
-            }
-        }
+        jntIndex = model.getJointIndex(v.name[i]);
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: You should check that the joint does not have 0 dof, otherwise the line
~~~
 m_jointPos(model.getJoint(jntIndex)->getDOFsOffset()) = v.position[i];
~~~
will be an invalid memory access.

**Comment by traversaro** [comment]
Travis failure are actually due to https://github.com/robotology/yarp/issues/2048 .

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp (hunk 9 above)
Comment (lrapetti) [reply src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: Thanks, fixed with https://github.com/robotology/idyntree/pull/538/commits/328233115dd4e937ec05bb5b6848cfdd6948c919.

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp [hunk 10]
```diff
@@ -19,7 +19,8 @@
 
 #include <yarp/math/Math.h>
 
-#include <iDynTree/Model/Model.h>
+// #include <iDynTree/Model/Model.h>
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]:
```suggestion
```
If that is not used, I suggest to just remove it.

Comment (lrapetti) [reply src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: Thanks, yes I forgot to remove it

**Comment by traversaro** [comment]
Do you want to cleanup the commits? Otherwise I can squash&merge.

**Comment by lrapetti** [comment]
> Do you want to cleanup the commits? Otherwise I can squash&merge.

I think we can just squash

**Comment by lrapetti** [comment]
CC @Yeshasvitvs  @diegoferigo

**Comment by traversaro** [comment]
Ack, I will wait for the lasted Travis&AppVeyor build, and then I will merge, thanks!

**Comment by traversaro** [comment]
Thanks!

**Comment by Yeshasvitvs** [comment]
@lrapetti have you tested the devel branch by running two yarprobotstatepublishers ?? I think the this PR doesn't add all the features present in the original branches. I tried today by running two instances of yarprobotstatepublishers but couldn't start them as there is a conflict with the client names.

**Comment by Yeshasvitvs** [comment]
My bad, the aliases used on icub30 doesn’t use the name-prefix and hence there was a problem in launching two yapreobotstatepublishers at the same time. However, the joints size check update is not present in the development and yarprobotstatepublisher could not be launched with the real robot as the fingers joints are not present in the robot urdf model.

**Comment by lrapetti** [comment]
Yhe 'yarprobotstatepublisher` with two human models in the `devel` branch has been extensively tested. However it has not been tested with the `icub` model, as discussed in previous treads the joint size check was still under discussion so it is not present in this PR. It should be decided how to handle this it and then open a PR
